components/raftstore/src/store/snap.rs

```diff
@@ -1966,6 +1990,66 @@ impl SnapManagerCore {
     }
 }
 
+/// `SnapRecvConcurrencyLimiter` enforces a limit on the number of simultaneous
+/// snapshot receives. It is consulted before a snapshot is generated.
+/// employs a TTL mechanism to automatically evict operations that have been
+/// pending longer than the specified TTL. The TTL helps to handle scenarios
+/// where a snapshot fails to be sent for any reason.
+#[derive(Clone)]
+pub struct SnapRecvConcurrencyLimiter {
+    limit: Arc<AtomicUsize>,
+    ttl_secs: u64,
+    timestamps: Arc<Mutex<VecDeque<Instant>>>,
+}
+
+impl SnapRecvConcurrencyLimiter {
+    pub fn new(limit: usize, ttl_secs: u64) -> Self {
+        SnapRecvConcurrencyLimiter {
+            limit: Arc::new(AtomicUsize::new(limit)),
+            ttl_secs,
+            timestamps: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    // Attempts to add a snapshot receive operation if below the concurrency
+    // limit. Returns true if the operation is allowed, false otherwise.
+    pub fn try_recv(&self) -> bool {
+        let limit = self.limit.load(Ordering::Relaxed);
+        if limit == 0 {
+            return true;
+        }
+        let mut timestamps = self.timestamps.lock().unwrap();
+        let current_time = Instant::now();
+        self.evict_expired_timestamps(&mut timestamps, current_time);
+
+        if timestamps.len() < limit {
+            timestamps.push_back(current_time);
+            return true;
+        }
+        false
+    }
+
+    fn evict_expired_timestamps(&self, timestamps: &mut VecDeque<Instant>, current_time: Instant) {
+        while let Some(&timestamp) = timestamps.front()
+            && current_time.duration_since(timestamp) > Duration::from_secs(self.ttl_secs)
+        {
+            timestamps.pop_front();
+        }
+    }
+
+    // Completes a snapshot receive operation by removing a timestamp from the
+    // queue. It is sufficient to remove the head of the queue instead of
+    // finding the matching timestamp, as we are only concerned with maintaining
+    // a total count of active operations.
+    pub fn finish_recv(&self) {
```

Comment:
Without matching timestamp, will `evict_expired_timestamps` leak operations?

```
initial state:                     queue: [ts1, ts2, ts3, ts5]
ts5 calls finish_recv():           queue: [ts2, ts3, ts5]
evict_expired_timestamps(ts4+ttl): queue: [ts5]
```
In this case, it leaks the operation `ts5`.

Reply:
It looks like you are describing an example where `finish_recv()` is never called on [ts1, ts2, ts3] and is only called on ts5. In your example, ideally the queue should be empty when `time > ts3+ttl`, but with the current implementation, the queue is not emptied until `time > ts5+ttl`. 

You can say that's a leak because it causes some delay in the entry's expiration, but I don't think it does much harm to the current purpose of the concurrency limiter. The delay in expiration simply makes the limiter more conservative for some time. Eventually all stale entries will be cleaned up after the TTL.